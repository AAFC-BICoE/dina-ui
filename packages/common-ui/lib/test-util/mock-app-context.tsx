import { mount } from "enzyme";
import { merge, noop } from "lodash";
import { useMemo, useRef } from "react";
import { SWRConfig } from "swr";
import { PartialDeep } from "type-fest";
import { AccountContextI, AccountProvider } from "../account/AccountProvider";
import {
  ApiClientI,
  ApiClientImpl,
  ApiClientProvider
} from "../api-client/ApiClientContext";
import { CommonUIIntlProvider } from "../intl/common-ui-intl";
import { ModalProvider } from "../modal/modal";

interface MockAppContextProviderProps {
  apiContext?: PartialDeep<ApiClientI>;
  accountContext?: Partial<AccountContextI>;
  children?: React.ReactNode;
}

/**
 * Wraps a test-rendered component to provide the contexts that would be available in
 * the application.
 */
export function MockAppContextProvider({
  accountContext,
  apiContext = { apiClient: { get: () => undefined as any } },
  children
}: MockAppContextProviderProps) {
  const DEFAULT_MOCK_ACCOUNT_CONTEXT: AccountContextI = useMemo(
    () => ({
      authenticated: true,
      groupNames: ["aafc", "cnc"],
      initialized: true,
      login: noop,
      logout: noop,
      roles: ["user"],
      token: "test-token",
      username: "test-user",
      isAdmin: false
    }),
    []
  );

  const DEFAULT_API_CONTEXT_VALUE = useMemo(
    () =>
      new ApiClientImpl({
        newId: () => "00000000-0000-0000-0000-000000000000"
      }),
    []
  );

  const apiContextWithWarnings = {
    ...apiContext,
    // Add a warning when bulkGet doesn't return anything in a test:
    bulkGet: async (paths: string[], options) => {
      const resources = await apiContext?.bulkGet?.(paths, options);
      if (!resources) {
        console.warn("No response returned for bulkGet paths: ", paths);
      } else {
        resources.forEach((resource, index) => {
          if (!resource) {
            console.warn("No value returned for bulkGet path: ", paths[index]);
          }
        });
      }
      return resources;
    }
  };

  const modalWrapperRef = useRef<HTMLDivElement>(null);

  return (
    <SWRConfig
      // Reset SWR cache between tests.
      value={{ provider: () => new Map() }}
    >
      <AccountProvider
        value={{ ...DEFAULT_MOCK_ACCOUNT_CONTEXT, ...accountContext }}
      >
        <ApiClientProvider
          value={merge({}, DEFAULT_API_CONTEXT_VALUE, apiContextWithWarnings)}
        >
          <CommonUIIntlProvider>
            <div ref={modalWrapperRef}>
              <ModalProvider appElement={modalWrapperRef.current}>
                {children}
              </ModalProvider>
            </div>
          </CommonUIIntlProvider>
        </ApiClientProvider>
      </AccountProvider>
    </SWRConfig>
  );
}

/**
 * Helper function to get a test wrapper with the required context providers.
 */
export function mountWithAppContext(
  element: React.ReactNode,
  mockAppContextProviderProps?: MockAppContextProviderProps
) {
  return mount(
    <MockAppContextProvider {...mockAppContextProviderProps}>
      {element}
    </MockAppContextProvider>
  );
}
