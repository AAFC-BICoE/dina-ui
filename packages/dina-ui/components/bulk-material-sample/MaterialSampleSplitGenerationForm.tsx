import { useEffect, useMemo, useState } from "react";
import { BackToListButton } from "common-ui/lib/button-bar/BackToListButton";
import PageLayout from "../page/PageLayout";
import {
  BackButton,
  DinaForm,
  NumberSpinnerField,
  SubmitButton,
  useApiClient,
  LoadingSpinner,
  DinaFormOnSubmit
} from "common-ui";
import { Card } from "react-bootstrap";
import { DinaMessage, useDinaIntl } from "../../intl/dina-ui-intl";
import { useFormikContext } from "formik";
import { MaterialSampleIdentifierGenerator } from "../../types/collection-api/resources/MaterialSampleIdentifierGenerator";
import { useBulkGet, useStringArrayConverter } from "common-ui";
import { MaterialSample } from "../../types/collection-api";
import { InputResource } from "kitsu";
import { SplitConfiguration } from "../../types/collection-api/resources/SplitConfiguration";
import { startCase } from "lodash";

const ENTITY_LINK = "/collection/material-sample";

interface MaterialSampleBulkSplitFields {
  numberToCreate: number;
}

interface MaterialSampleSplitGenerationFormProps {
  ids: string[];
  splitConfiguration?: SplitConfiguration;
  onGenerate: (samples: InputResource<MaterialSample>[]) => void;
}

export function MaterialSampleSplitGenerationForm({
  ids,
  splitConfiguration,
  onGenerate
}: MaterialSampleSplitGenerationFormProps) {
  const { formatMessage } = useDinaIntl();
  const [convertArrayToString] = useStringArrayConverter();

  const isMultiple = useMemo(() => ids.length > 1, [ids]);

  const [generatedIdentifiers, setGeneratedIdentifiers] = useState<string[]>(
    []
  );

  const splitFromMaterialSamples = useBulkGet<MaterialSample>({
    ids,
    listPath:
      "collection-api/material-sample?include=materialSampleChildren,collection,parentMaterialSample",
    disabled: ids.length === 0
  });

  const buttonBar = (
    <>
      {/* Back Button (Changes depending on the number of records) */}
      {isMultiple ? (
        <BackToListButton entityLink={ENTITY_LINK} />
      ) : (
        <BackButton entityLink={ENTITY_LINK} entityId={ids[0]} />
      )}

      {/* Submit Button */}
      <SubmitButton className={"ms-auto"}>
        <DinaMessage id="splitButton" />
      </SubmitButton>
    </>
  );

  if (splitFromMaterialSamples.loading) {
    return <LoadingSpinner loading={true} />;
  }

  const initialValues: MaterialSampleBulkSplitFields = {
    numberToCreate: 1
  };

  const onSubmit: DinaFormOnSubmit<MaterialSampleBulkSplitFields> = ({
    submittedValues
  }) => {
    if (
      Number(generatedIdentifiers.length) !==
      Number(submittedValues?.numberToCreate)
    ) {
      return;
    }

    const splitFromMaterialSample: any = splitFromMaterialSamples?.data?.[0];

    const samples: InputResource<MaterialSample>[] = [];
    generatedIdentifiers.forEach((identifier) => {
      samples.push({
        type: "material-sample",
        parentMaterialSample: {
          id: splitFromMaterialSample?.id ?? "",
          type: "material-sample"
        },
        group: splitFromMaterialSample?.group ?? "",
        collection: splitFromMaterialSample?.collection?.id
          ? {
              id: splitFromMaterialSample?.collection?.id ?? "",
              type: "collection"
            }
          : undefined,
        publiclyReleasable: true,
        allowDuplicateName: false,
        materialSampleName: identifier
      });
    });

    onGenerate(samples);
  };

  return (
    <DinaForm<MaterialSampleBulkSplitFields>
      initialValues={initialValues}
      onSubmit={onSubmit}
    >
      <PageLayout titleId="splitSubsampleTitle" buttonBarContent={buttonBar}>
        <div className="row">
          <div className="col-md-5">
            <h4 className="mt-2">
              <DinaMessage id="settingLabel" />
            </h4>
            <Card>
              <Card.Body>
                <DinaMessage id="splitFrom" />:
                <span className="ms-2">
                  {convertArrayToString(
                    (splitFromMaterialSamples?.data as any)?.map(
                      (materialSample) => materialSample?.materialSampleName
                    )
                  )}
                </span>
              </Card.Body>
            </Card>

            <NumberSpinnerField
              name="numberToCreate"
              min={1}
              max={500}
              label={formatMessage("materialSamplesToCreate")}
              disabled={isMultiple}
              className="mt-3"
            />
          </div>
          <div className="col-md-7">
            <PreviewGeneratedNames
              splitFromMaterialSamples={
                splitFromMaterialSamples.data as MaterialSample[]
              }
              generatedIdentifiers={generatedIdentifiers}
              setGeneratedIdentifiers={setGeneratedIdentifiers}
              splitConfiguration={splitConfiguration}
            />
          </div>
        </div>
      </PageLayout>
    </DinaForm>
  );
}

interface PreviewGeneratedNamesProps {
  splitFromMaterialSamples: MaterialSample[];
  generatedIdentifiers: string[];
  setGeneratedIdentifiers: (identifiers: string[]) => void;
  splitConfiguration?: SplitConfiguration;
}

function PreviewGeneratedNames({
  splitFromMaterialSamples,
  generatedIdentifiers,
  setGeneratedIdentifiers,
  splitConfiguration
}: PreviewGeneratedNamesProps) {
  const { save } = useApiClient();
  const formik = useFormikContext<MaterialSampleBulkSplitFields>();

  const numberToCreate = formik.values.numberToCreate;

  function getIdentifierRequest(index): MaterialSampleIdentifierGenerator {
    return {
      type: "material-sample-identifier-generator",
      amount: numberToCreate,
      currentParentUUID: splitFromMaterialSamples?.[index]?.id ?? "",
      strategy:
        splitConfiguration?.materialSampleNameGeneration?.strategy ??
        "DIRECT_PARENT",
      characterType:
        splitConfiguration?.materialSampleNameGeneration?.characterType ??
        "LOWER_LETTER",
      materialSampleType:
        splitConfiguration?.materialSampleNameGeneration?.materialSampleType
    };
  }

  // To prevent spamming the network calls, this useEffect has a debounce.
  useEffect(() => {
    async function callGenerateIdentifierAPI() {
      const response = await save<MaterialSampleIdentifierGenerator>(
        [
          {
            resource: getIdentifierRequest(0),
            type: "material-sample-identifier-generator"
          }
        ],
        { apiBaseUrl: "/collection-api", overridePatchOperation: true }
      );

      const generatedIdentifiersResults = response.flatMap(
        (resp) => resp?.nextIdentifiers ?? []
      );

      setGeneratedIdentifiers(generatedIdentifiersResults);
    }

    let timeoutId;
    const debouncedHandleChange = () => {
      clearTimeout(timeoutId);
      timeoutId = setTimeout(() => {
        callGenerateIdentifierAPI();
      }, 500);
    };
    debouncedHandleChange();
    return () => {
      clearTimeout(timeoutId);
    };
  }, [formik.values]);

  // Columns to be displayed
  const materialSampleType =
    splitConfiguration?.materialSampleNameGeneration?.materialSampleType ?? "";
  const formattedMaterialSampleType = startCase(
    materialSampleType.toLowerCase().replace(/_/g, " ")
  );

  return (
    <div className="mt-2">
      <h4>
        <DinaMessage id="previewLabel" />
      </h4>
      <table className="table">
        <thead>
          <tr>
            <th>
              <DinaMessage id="splitPreviewNumberColumn" />
            </th>
            <th>
              <DinaMessage id="splitPreviewGeneratedIdentifierColumn" />
            </th>
            <th>
              <DinaMessage id="parentMaterialSample" />
            </th>
            <th>
              <DinaMessage id="field_materialSampleType" />
            </th>
          </tr>
        </thead>
        <tbody>
          {Array.from(
            {
              length: numberToCreate
            },
            (_, i) => i
          ).map((_, index) => (
            <tr key={index + 1}>
              <td>#{index + 1}</td>
              <td>
                {generatedIdentifiers[index] ? (
                  generatedIdentifiers[index]
                ) : (
                  <LoadingSpinner loading={true} />
                )}
              </td>
              <td>{splitFromMaterialSamples?.[0]?.materialSampleName ?? ""}</td>
              <td>{formattedMaterialSampleType}</td>
            </tr>
          ))}
        </tbody>
      </table>
    </div>
  );
}
