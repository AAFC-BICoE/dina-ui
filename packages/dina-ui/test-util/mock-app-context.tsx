import {
  AccountContextI,
  AccountProvider,
  ApiClientI,
  ApiClientImpl,
  ApiClientProvider,
  ModalProvider
} from "common-ui";
import { mount } from "enzyme";
import { merge, noop } from "lodash";
import { useMemo, useRef } from "react";
import { DndProvider } from "react-dnd-cjs";
import HTML5Backend from "react-dnd-html5-backend-cjs";
import { SWRConfig } from "swr";
import { PartialDeep } from "type-fest";
import { FileUploadProviderImpl } from "../components/object-store/file-upload/FileUploadProvider";
import { DinaIntlProvider } from "../intl/dina-ui-intl";

interface MockAppContextProviderProps {
  apiContext?: PartialDeep<ApiClientI>;
  accountContext?: Partial<AccountContextI>;
  children?: React.ReactNode;
}

/**
 * Wraps a test-rendered component to provide the contexts that would be available in
 * the application.
 */
export function MockAppContextProvider({
  accountContext,
  apiContext = { apiClient: { get: () => undefined as any } },
  children
}: MockAppContextProviderProps) {
  const DEFAULT_MOCK_ACCOUNT_CONTEXT: AccountContextI = useMemo(
    () => ({
      authenticated: true,
      groupNames: ["aafc", "cnc"],
      initialized: true,
      login: noop,
      logout: noop,
      roles: ["user"],
      token: "test-token",
      username: "test-user",
      isAdmin: false
    }),
    []
  );

  const DEFAULT_API_CONTEXT_VALUE = useMemo(
    () =>
      new ApiClientImpl({
        newId: () => "00000000-0000-0000-0000-000000000000"
      }),
    []
  );

  const apiContextWithWarnings = {
    ...apiContext,
    // Add a warning when bulkGet doesn't return anything in a test:
    bulkGet: async (paths: string[], options) => {
      const resources = await apiContext?.bulkGet?.(paths, options);
      if (!resources) {
        console.warn("No response returned for bulkGet paths: ", paths);
      } else {
        resources.forEach((resource, index) => {
          if (!resource) {
            console.warn("No value returned for bulkGet path: ", paths[index]);
          }
        });
      }
      return resources;
    }
  };

  const modalWrapperRef = useRef<HTMLDivElement>(null);

  return (
    <SWRConfig
      // Reset SWR cache between tests.
      value={{ provider: () => new Map() }}
    >
      <AccountProvider
        value={{ ...DEFAULT_MOCK_ACCOUNT_CONTEXT, ...accountContext }}
      >
        <ApiClientProvider
          value={merge({}, DEFAULT_API_CONTEXT_VALUE, apiContextWithWarnings)}
        >
          <FileUploadProviderImpl>
            <DinaIntlProvider>
              <DndProvider backend={HTML5Backend}>
                <div ref={modalWrapperRef}>
                  <ModalProvider appElement={modalWrapperRef.current}>
                    {children}
                  </ModalProvider>
                </div>
              </DndProvider>
            </DinaIntlProvider>
          </FileUploadProviderImpl>
        </ApiClientProvider>
      </AccountProvider>
    </SWRConfig>
  );
}

/**
 * Helper function to get a test wrapper with the required context providers.
 */
export function mountWithAppContext(
  element: React.ReactNode,
  mockAppContextProviderProps?: MockAppContextProviderProps
) {
  return mount(
    <MockAppContextProvider {...mockAppContextProviderProps}>
      {element}
    </MockAppContextProvider>
  );
}
