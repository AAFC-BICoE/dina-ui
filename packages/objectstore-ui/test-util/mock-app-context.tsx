import {
  AccountContextI,
  AccountProvider,
  ApiClientContextI,
  AuthenticatedApiClientProvider,
  createContextValue,
  ModalProvider
} from "common-ui";
import { mount } from "enzyme";
import { merge, noop } from "lodash";
import { ObjectStoreIntlProvider } from "../intl/objectstore-intl";

interface MockAppContextProviderProps {
  apiContext?: Partial<ApiClientContextI>;
  accountContext?: Partial<AccountContextI>;
  children?: React.ReactNode;
}

/**
 * Wraps a test-rendered component to provide the contexts that would be available in
 * the application.
 */
export function MockAppContextProvider({
  accountContext,
  apiContext,
  children
}: MockAppContextProviderProps) {
  return (
    <AccountProvider
      value={{ ...DEFAULT_MOCK_ACCOUNT_CONTEXT, ...accountContext }}
    >
      <AuthenticatedApiClientProvider
        apiContext={merge({}, DEFAULT_API_CONTEXT_VALUE, apiContext)}
      >
        <ObjectStoreIntlProvider>
          <ModalProvider appElement={document.querySelector("body")}>
            {children}
          </ModalProvider>
        </ObjectStoreIntlProvider>
      </AuthenticatedApiClientProvider>
    </AccountProvider>
  );
}

/**
 * Helper function to get a test wrapper with the required context providers.
 */
export function mountWithAppContext(
  element: React.ReactNode,
  mockAppContextProviderProps?: MockAppContextProviderProps
) {
  return mount(
    <MockAppContextProvider {...mockAppContextProviderProps}>
      {element}
    </MockAppContextProvider>
  );
}

const DEFAULT_MOCK_ACCOUNT_CONTEXT: AccountContextI = {
  authenticated: true,
  groups: ["/aafc", "cnc"],
  initialized: true,
  login: noop,
  logout: noop,
  token: "test-token",
  username: "test-user"
};

const DEFAULT_API_CONTEXT_VALUE = createContextValue({
  getTempIdGenerator: () => () => "00000000-0000-0000-0000-000000000000"
});
