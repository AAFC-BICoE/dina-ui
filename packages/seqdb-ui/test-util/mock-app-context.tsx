import {
  ApiClientContext,
  ApiClientContextI,
  createContextValue
} from "common-ui";
import { mount } from "enzyme";
import { SeqdbIntlProvider } from "../intl/seqdb-intl";

interface MockAppContextProviderProps {
  apiContext?: ApiClientContextI;
  children?: React.ReactNode;
}

/**
 * Wraps a test-rendered component to provide the contexts that would be available in
 * the application.
 */
export function MockAppContextProvider({
  apiContext,
  children
}: MockAppContextProviderProps) {
  return (
    <ApiClientContext.Provider value={apiContext || createContextValue()}>
      <SeqdbIntlProvider>{children}</SeqdbIntlProvider>
    </ApiClientContext.Provider>
  );
}

/**
 * Helper function to get a test wrapper with the required context providers.
 */
export function mountWithAppContext(
  element: React.ReactNode,
  mockAppContextProviderProps?: MockAppContextProviderProps
) {
  return mount(
    <MockAppContextProvider {...mockAppContextProviderProps}>
      {element}
    </MockAppContextProvider>
  );
}
