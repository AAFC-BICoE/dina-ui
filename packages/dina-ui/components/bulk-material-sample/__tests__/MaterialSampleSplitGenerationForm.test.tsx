import { mountWithAppContext } from "../../../test-util/mock-app-context";
import { MaterialSampleSplitGenerationForm } from "../MaterialSampleSplitGenerationForm";

const NO_CHILDREN_MATERIAL_SAMPLE_UUID = "a503d31d-8203-4766-af85-db271e087853";

const NO_CHILDREN_MATERIAL_SAMPLE = {
  id: NO_CHILDREN_MATERIAL_SAMPLE_UUID,
  type: "material-sample",
  group: "aafc",
  materialSampleName: "CNC-1",
  materialSampleChildren: [],
  parentMaterialSample: undefined
};

const CHILDREN_MATERIAL_SAMPLE_UUID = "3a0ec067-6a98-4dba-8866-443237483a56";

const CHILDREN_MATERIAL_SAMPLE = {
  id: NO_CHILDREN_MATERIAL_SAMPLE_UUID,
  type: "material-sample",
  group: "aafc",
  materialSampleName: "CNC-2",
  materialSampleChildren: [
    {
      ordinal: 0,
      materialSampleName: "CNC-2-a"
    },
    {
      ordinal: 1,
      materialSampleName: "CNC-2-b"
    },
    {
      ordinal: 2,
      materialSampleName: "CNC-2-c"
    }
  ]
};

const mockOnGenerate = jest.fn();

const mockBulkGet = jest.fn<any, any>(async (paths: string[]) =>
  paths.map((path) => {
    switch (path) {
      case `material-sample/${NO_CHILDREN_MATERIAL_SAMPLE_UUID}?include=materialSampleChildren,collection`:
        return NO_CHILDREN_MATERIAL_SAMPLE;
      case `material-sample/${CHILDREN_MATERIAL_SAMPLE_UUID}?include=materialSampleChildren,collection`:
        return CHILDREN_MATERIAL_SAMPLE;
      default:
        return {};
    }
  })
);

const mockSave = jest.fn();

const apiContext: any = { bulkGet: mockBulkGet, save: mockSave };

describe("MaterialSampleSplitGenerationForm", () => {
  beforeEach(() => jest.clearAllMocks);

  it("Layout snapshot", async () => {
    const wrapper = mountWithAppContext(
      <MaterialSampleSplitGenerationForm
        onGenerate={mockOnGenerate}
        ids={[NO_CHILDREN_MATERIAL_SAMPLE_UUID]}
      />,
      { apiContext }
    );

    // Wait for the material sample request.
    await new Promise(setImmediate);
    wrapper.update();

    // This test is designed to simply ensure the design doesn't change when it's not expected.
    // If it's expected, the snapshot will need to be updated.
    expect(wrapper.find("main").debug()).toMatchSnapshot();
  });
});
