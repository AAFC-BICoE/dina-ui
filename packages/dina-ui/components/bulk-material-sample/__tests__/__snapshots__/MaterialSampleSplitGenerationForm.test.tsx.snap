// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MaterialSampleSplitGenerationForm Layout snapshot 1`] = `
"<main className=\\"container-fluid px-5\\" role=\\"main\\">
  <h1 id=\\"wb-cont\\">
    Split Material Sample
  </h1>
  <div className=\\"row\\">
    <div className=\\"col-md-5\\">
      <h4 className=\\"mt-2\\">
        <FormattedMessage id=\\"settingLabel\\">
          <FormattedMessage defaultMessage=\\" \\" id=\\"settingLabel\\">
            Settings
          </FormattedMessage>
        </FormattedMessage>
      </h4>
      <Card body={false}>
        <div className=\\"card\\">
          <CardBody>
            <div className=\\"card-body\\">
              <FormattedMessage id=\\"splitFrom\\">
                <FormattedMessage defaultMessage=\\" \\" id=\\"splitFrom\\">
                  Split from
                </FormattedMessage>
              </FormattedMessage>
              :
              <span className=\\"ms-2\\" />
            </div>
          </CardBody>
        </div>
      </Card>
      <NumberSpinnerField name=\\"numberToCreate\\" min={1} max={500} label=\\"Material Samples to Create\\" disabled={false} className=\\"mt-3\\">
        <FieldWrapper name=\\"numberToCreate\\" min={1} max={500} label=\\"Material Samples to Create\\" disabled={false} className=\\"mt-3\\">
          <FieldSpy fieldName=\\"numberToCreate\\" validate={[undefined]}>
            <FormikConnect(FastFieldInner) name=\\"numberToCreate\\" shouldUpdate={[Function: shouldRender]} validate={[undefined]}>
              <FastFieldInner name=\\"numberToCreate\\" shouldUpdate={[Function: shouldRender]} validate={[undefined]} formik={{...}}>
                <FieldSpyInternal fastFieldProps={{...}} fieldName=\\"numberToCreate\\">
                  <LabelWrapper fieldWrapperProps={{...}} fieldSpyProps={{...}}>
                    <div className=\\"mt-3\\">
                      <label className=\\"numberToCreate-field mb-2 w-100 mb-3\\" htmlFor={[undefined]}>
                        <div className=\\"field-label mb-2\\">
                          <strong>
                            Material Samples to Create
                          </strong>
                        </div>
                        <div className={[undefined]} style={{...}}>
                          <FormikConnectedField fieldSpyProps={{...}} fieldWrapperProps={{...}}>
                            <ErrorBoundary FallbackComponent={[Function: FallbackComponent]}>
                              <input className=\\"form-control\\" type=\\"number\\" min={1} max={500} size={4} step={1} onKeyDown={[Function: onKeyDown]} onChange={[Function: onChange]} onClick={[Function: onClick]} disabled={false} value={1} />
                            </ErrorBoundary>
                          </FormikConnectedField>
                        </div>
                      </label>
                    </div>
                  </LabelWrapper>
                </FieldSpyInternal>
              </FastFieldInner>
            </FormikConnect(FastFieldInner)>
          </FieldSpy>
        </FieldWrapper>
      </NumberSpinnerField>
    </div>
    <div className=\\"col-md-7\\">
      <PreviewGeneratedNames splitFromMaterialSamples={{...}} generatedIdentifiers={{...}} setGeneratedIdentifiers={[Function: bound dispatchAction]} splitConfiguration={[undefined]}>
        <div className=\\"mt-2\\">
          <h4>
            <FormattedMessage id=\\"previewLabel\\">
              <FormattedMessage defaultMessage=\\" \\" id=\\"previewLabel\\">
                Preview
              </FormattedMessage>
            </FormattedMessage>
          </h4>
          <table className=\\"table\\">
            <thead>
              <tr>
                <th>
                  <FormattedMessage id=\\"splitPreviewNumberColumn\\">
                    <FormattedMessage defaultMessage=\\" \\" id=\\"splitPreviewNumberColumn\\">
                      Number
                    </FormattedMessage>
                  </FormattedMessage>
                </th>
                <th>
                  <FormattedMessage id=\\"splitPreviewGeneratedIdentifierColumn\\">
                    <FormattedMessage defaultMessage=\\" \\" id=\\"splitPreviewGeneratedIdentifierColumn\\">
                      Generated Identifier
                    </FormattedMessage>
                  </FormattedMessage>
                </th>
                <th>
                  <FormattedMessage id=\\"parentMaterialSample\\">
                    <FormattedMessage defaultMessage=\\" \\" id=\\"parentMaterialSample\\">
                       Parent Material Sample
                    </FormattedMessage>
                  </FormattedMessage>
                </th>
                <th>
                  <FormattedMessage id=\\"field_materialSampleType\\">
                    <FormattedMessage defaultMessage=\\" \\" id=\\"field_materialSampleType\\">
                      Material Sample Type
                    </FormattedMessage>
                  </FormattedMessage>
                </th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <td>
                  #
                  1
                </td>
                <td>
                  <LoadingSpinner loading={true}>
                    <div className=\\"spinner-border\\" role=\\"status\\">
                      <span className=\\"visually-hidden mb-3\\">
                        Loading...
                      </span>
                    </div>
                  </LoadingSpinner>
                </td>
                <td />
                <td />
              </tr>
            </tbody>
          </table>
        </div>
      </PreviewGeneratedNames>
    </div>
  </div>
</main>"
`;
